### ISSUE TITLE
[IDEA] 'Neglected' module

### ISSUE THREAD

**Issue by BWPanda** [backdrop-ops/contrib#375]
I just had an idea for a new module: 'Neglected'.

Neglected would tell you when there are modules enabled on your site that aren't being used. A report would be shown in the Reports section listing such modules, and a notification displayed on the Status page if such modules exist.

Here's a use-case:
You install the [Telephone](https://github.com/backdrop-contrib/telephone) module and add a Telephone field to a content type. Later, you realise you don't need it and so delete the field. But you forget to uninstall the module, so it stays there taking up space, possibly requiring security updates, etc.

With this module, you'd see a notification in the Status Report showing neglected modules. When you visit the Neglected report page you'll see the Telephone module listed, for example, (and any others) and can then decide if you want to utilise it or uninstall/disable it.

The 'neglected' state of a module would be determined via a hook that modules can implement. Core modules would be accounted for in Neglected's code itself. For example, the Telephone module would implement the hook and return TRUE if there were no fields of type 'Telephone' on the site. The core Contact module's implementation (in Neglected's code) would return TRUE if there are no contact form categories. These hooks would be called on cron runs and update the report accordingly.

If I get around to it I might try my hand at writing this module. And it only just occurs to me now (after writing all of the above) to check if such a module already exists in Drupal... After searching, I found two similar modules:

- [Unused Modules](https://www.drupal.org/project/unused_modules) - this is similar in concept, but also different in a few aspects. Perhaps a port of that module with some additions from my idea above could work nicely...
- [Module analyzer](https://www.drupal.org/project/module_analyzer) - seems to take a different approach and makes a lot of assumptions...

It also just occurs to me now that this module would only be useful if it checked all the modules on your site, but it wouldn't be likely that all contrib modules would want to add a new hook implementation just for this module to work... So maybe this isn't a good idea after all. Oh well, maybe it'll inspire someone else with a better idea instead :wink:

**Comment by jenlampton** [comment]
I love this idea, but I'm not sure about the name. "Neglected" sounds to me like the module suffering in some way, or hasn't been taken care of (pudding an undue pressure on maintiners, maybe?)

What about something like "underutilized"? 

Naming things is hard...

**Comment by stpaultim** [comment]
My first response to the name was also that the module was not being taken care of the maintainers. Even after reading "Neglected would tell you when there are modules enabled on your site that aren't being used," I was thinking modules that we're getting used by other users. Not modules that are not being used on your site. 

I would like a module like this, but am in favor of another name. 

Brainstorming ideas....
* Inactive modules
* Housekeeping
* Module monitor
* Sleeper peeper 
* Gone fishing
* Overlooked
* Forget me not

**Comment by BWPanda** [comment]
I think the name's not that important. I'd likely just call it 'Unused' then.

But I'd really like feedback on the possible issue mentioned in the last paragraph:

> It also just occurs to me now that this module would only be useful if it checked all the modules on your site, but it wouldn't be likely that all contrib modules would want to add a new hook implementation just for this module to work...

Is this even going to be possible/ worth it then?

**Comment by stpaultim** [comment]
Naming is hard, but it can be fun too. ;-) But, I agree it's not that important. 

I would agree, that if this module would require that all other modules implement code to make it work, that would be a problem. I find it pretty unlikely that other modules are going to find the time to do this, UNLESS you want to go through all the modules and file PR's. (Probably not worth it).

I think this module might be worth it if it works on 80-90% of the top 100 modules. But if it's less than that, it's worth might be questionable.

**Comment by jenlampton** [comment]
I wonder if we can use a few of the most common hooks instead of adding a
new one? Most modules implement hook_menu() and/or hook_config_info(). What
we're you thinking of checking, and can it be done there?

Jen

**Comment by BWPanda** [comment]
Well we'd need a way to differentiate between an installed module that's in use, and an installed module that's not. The only way I can think of to do that is specific to each module.

For example, a module that provides a field would be considered 'in use' if a field of that type exists. If there are no fields of that type, then the module is clearly not being used. However, that's different from, say, a module that provides a block. If the block is present in a layout, then it's in use, otherwise it's not. So each module would need to return TRUE or FALSE depending on a check specific to it.

I don't think this can be done with existing hooks...

**Comment by jenlampton** [comment]
> I don't think this can be done with existing hooks...

Yeah, you're right. I went down the same train-of-thought after I posted the comment. Even if we could put the test into an existing hook, the tests would need to be completely different. And what if all the module did was alter a form? 

> Is this even going to be possible/ worth it then?

I still think it would be worth it. I could see this module getting put on a list for inclusion in core. The goal for this module completely matches what we want Backdrop the product to do: to help less experienced people build better sites. Those people will have the least ability to determine whether a module is in-use or not, but the contrib developers know exactly what to test. 

Because of this, I'd be willing to add support for this new hook in all my contrib projects. :)

**Comment by BWPanda** [comment]
I opened a feature request in core for this idea. Let's see what happens there...

https://github.com/backdrop/backdrop-issues/issues/5020

**Comment by jenlampton** [comment]
This would work nicely as a contrib module. Can we build it and see how
much it's used before adding something like this to core?

I suspect there are some modules that won't have a clear indicator of
whether they're being used or not, something like chosen, for example,
couldn't tell.

Jen

**Comment by BWPanda** [comment]
Here's my initial (working) proof-of-concept: https://github.com/backdrop-contrib/unused
